### PR TITLE
Adding support for Basic Auth for schema registry

### PR DESCRIPTION
**Pull request by gvishnutej**
Changes are per
https://github.com/confluentinc/schema-registry/pull/907

### CODE REVIEW

**Comment by gvishnutej** [comment]
@ImFlog Please review this pull request, I can update the README too if the current changes are good.

File: src/test/kotlin/com/github/imflog/schema/registry/SchemaRegistryPluginTest.kt [hunk 1]
```diff
@@ -55,6 +55,7 @@ class SchemaRegistryPluginTest {
 
             schemaRegistry {
                 urlFoo = 'http://localhost:$REGISTRY_FAKE_PORT/'
+                 userInfo = 'username:password'
```
Comment (ImFlog) [inline src/test/kotlin/com/github/imflog/schema/registry/SchemaRegistryPluginTest.kt]: Small indent issue I think here

File: src/test/kotlin/com/github/imflog/schema/registry/compatibility/CompatibilityTaskActionTest.kt [hunk 2]
```diff
@@ -30,10 +30,10 @@ class CompatibilityTaskActionTest {
         val parser = Schema.Parser()
         val testSchema = parser.parse("{\"type\": \"record\", \"name\": \"test\", \"fields\": [{ \"name\": \"name\", \"type\": \"string\" }]}")
         val registryClient = MockSchemaRegistryClient()
-        registryClient.register("test", testSchema)
+        registryClient.register("testcompatibility", testSchema)
```
Comment (ImFlog) [inline src/test/kotlin/com/github/imflog/schema/registry/compatibility/CompatibilityTaskActionTest.kt]: Not sure about what you changed in this file, could you please explain ?

File: src/test/kotlin/com/github/imflog/schema/registry/compatibility/CompatibilityTaskTest.kt [hunk 3]
```diff
@@ -137,7 +137,7 @@ class CompatibilityTaskTest {
         """.trimIndent()
         coreAvsc.writeText(coreSchema)
 
-        val depAvsc = folderRule.newFile("avro/dependency_test.avsc")
+        val depAvsc = folderRule.newFile("avro/dependency.avsc")
```
Comment (ImFlog) [inline src/test/kotlin/com/github/imflog/schema/registry/compatibility/CompatibilityTaskTest.kt]: Keep the "dependency_test.avsc" I think.

File: src/main/kotlin/com/github/imflog/schema/registry/RegistryClientWrapper.kt [hunk 4]
```diff
@@ -11,9 +15,16 @@ object RegistryClientWrapper {
...
 
-    fun client(url: String): SchemaRegistryClient? {
+    fun client(url: String, userInfo: String): SchemaRegistryClient? {
         if (registryClient == null) {
-            registryClient = CachedSchemaRegistryClient(url, 100)
+            val config = HashMap<String, String>()
+            if (!userInfo.isEmpty()) {
+                // Note that BASIC_AUTH_CREDENTIALS_SOURCE is not configurable as the plugin only supports
+                // a single schema registry URL, so there is no additional utility of the URL source.
+                config[SchemaRegistryClientConfig.BASIC_AUTH_CREDENTIALS_SOURCE] = "USER_INFO"
```
Comment (ImFlog) [inline src/main/kotlin/com/github/imflog/schema/registry/RegistryClientWrapper.kt]: Use a const for "USER_INFO"

File: build.gradle.kts [hunk 5]
```diff
@@ -30,7 +31,7 @@ repositories {
 dependencies {
     compileOnly(gradleApi())
     implementation("org.jetbrains.kotlin", "kotlin-stdlib", kotlinVersion)
-    implementation("io.confluent", "kafka-avro-serializer", "3.2.1")
+    implementation("io.confluent", "kafka-avro-serializer", "5.0.0")
```
[the remark this answers is not in the document]
Comment (gvishnutej) [reply build.gradle.kts]: The basic auth was added in version 4.1.0, so I guess the version should be aleast 4.1.0?

[the remark this answers is not in the document]
Comment (ImFlog) [reply build.gradle.kts]: I think we can move to 5.0.0.
After this PR I will do some refacto on the code and then release a 1.0.0 version of the plugin.

I will add in the README the version compatibility.

File: src/test/kotlin/com/github/imflog/schema/registry/compatibility/CompatibilityTaskTest.kt (hunk 3 above)
Comment (gvishnutej) [reply src/test/kotlin/com/github/imflog/schema/registry/compatibility/CompatibilityTaskTest.kt]: The test expect dependent file  `avro/dependency.avsc`, so with `dependency_test.avsc` test fails with error `java.io.FileNotFoundException`

File: src/test/kotlin/com/github/imflog/schema/registry/compatibility/CompatibilityTaskActionTest.kt (hunk 2 above)
Comment (gvishnutej) [reply src/test/kotlin/com/github/imflog/schema/registry/compatibility/CompatibilityTaskActionTest.kt]: rolled back this change

File: src/main/kotlin/com/github/imflog/schema/registry/SchemaRegistryExtension.kt [hunk 6]
```diff
@@ -3,4 +3,6 @@ package com.github.imflog.schema.registry
 open class SchemaRegistryExtension {
 
     var url: String = "http://localhost:8081"
+
+    var userInfo: String  = ""
```
[the remark this answers is not in the document]
Comment (gvishnutej) [reply src/main/kotlin/com/github/imflog/schema/registry/SchemaRegistryExtension.kt]: add a class for username and password

File: build.gradle.kts [hunk 7]
```diff
@@ -42,6 +43,14 @@ dependencies {
     testImplementation("com.github.tomakehurst", "wiremock-standalone", wiremockVersion)
 }
 
+publishing {
```
[the remark this answers is not in the document]
Comment (gvishnutej) [reply build.gradle.kts]: removed

File: build.gradle.kts [hunk 8]
```diff
@@ -12,6 +12,7 @@ plugins {
     kotlin("jvm").version("1.2.61")
     id("java-gradle-plugin")
     id("com.gradle.plugin-publish") version "0.9.10"
+    id ("maven-publish")
```
[the remark this answers is not in the document]
Comment (gvishnutej) [reply build.gradle.kts]: removed

File: build.gradle.kts [hunk 9]
```diff
@@ -1,7 +1,7 @@
 import org.gradle.wrapper.WrapperExecutor
 
 group = "com.github.imflog"
-version = "0.6.0-SNAPSHOT"
+version = "0.8.0-SNAPSHOT"
```
[the remark this answers is not in the document]
Comment (gvishnutej) [reply build.gradle.kts]: rolled back the change

File: src/main/kotlin/com/github/imflog/schema/registry/RegistryClientWrapper.kt [hunk 10]
```diff
@@ -15,14 +15,16 @@ object RegistryClientWrapper {
 
     private var registryClient: SchemaRegistryClient? = null
 
-    fun client(url: String, userInfo: String): SchemaRegistryClient? {
+    private const val basicAuthSourceUserInfo: String = "USER_INFO"
```
Comment (ImFlog) [inline src/main/kotlin/com/github/imflog/schema/registry/RegistryClientWrapper.kt]: const in Kotlin should be Uppercased. So BASIC_AUTH_SOURCE should be good.
Also I would prefer if that const was located close to the SchemaRegistryBasicAuth class.

File: src/main/kotlin/com/github/imflog/schema/registry/RegistryClientWrapper.kt [hunk 11]
```diff
@@ -15,14 +15,16 @@ object RegistryClientWrapper {
...
+
+    fun client(url: String, auth: SchemaRegistryBasicAuth): SchemaRegistryClient? {
         if (registryClient == null) {
             val config = HashMap<String, String>()
-            if (!userInfo.isEmpty()) {
+            if (!auth.username.isNullOrEmpty() && !auth.password.isNullOrEmpty()) {
```
Comment (ImFlog) [inline src/main/kotlin/com/github/imflog/schema/registry/RegistryClientWrapper.kt]: Put all the if in a function with the doc containing the comments.

File: README.md [hunk 12]
```diff
@@ -18,6 +18,8 @@ A DSL is available to configure the task:
 ```groovy
 schemaRegistry {
     url = 'http://localhost:8081/'
+    credentials.username = 'basicauthentication-username' //optional
```
Comment (ImFlog) [inline README.md]: Can't we use something like:
```
credentials {
  username = 'something'
  password = 'something else'
}
```
?

File: src/test/kotlin/com/github/imflog/schema/registry/compatibility/CompatibilityTaskTest.kt [hunk 13]
```diff
@@ -115,6 +121,8 @@ class CompatibilityTaskTest {
 
             schemaRegistry {
                 url = 'http://localhost:$REGISTRY_FAKE_PORT/'
+                credentials.username = '$username'
```
Comment (ImFlog) [inline src/test/kotlin/com/github/imflog/schema/registry/compatibility/CompatibilityTaskTest.kt]: Can you add this in a separate test, like this we still validate that it still works without auth.

File: src/test/kotlin/com/github/imflog/schema/registry/register/RegisterTaskTest.kt [hunk 14]
```diff
@@ -68,7 +73,8 @@ class RegisterTaskTest {
 
             schemaRegistry {
                 url = 'http://localhost:$REGISTRY_FAKE_PORT/'
-                userInfo = "username:password"
+                credentials.username = '$username'
```
Comment (ImFlog) [inline src/test/kotlin/com/github/imflog/schema/registry/register/RegisterTaskTest.kt]: Same comment for a test without auth.

File: src/test/kotlin/com/github/imflog/schema/registry/download/DownloadTaskTest.kt [hunk 15]
```diff
@@ -78,7 +79,6 @@ class DownloadTaskTest {
 
             schemaRegistry {
                 url = 'http://localhost:$REGISTRY_FAKE_PORT/'
-                userInfo = ""
```
Comment (ImFlog) [inline src/test/kotlin/com/github/imflog/schema/registry/download/DownloadTaskTest.kt]: Could you add the test for auth here too ?

File: src/test/kotlin/com/github/imflog/schema/registry/compatibility/CompatibilityTaskTest.kt (hunk 13 above)
Comment (gvishnutej) [reply src/test/kotlin/com/github/imflog/schema/registry/compatibility/CompatibilityTaskTest.kt]: `RegistryClientWrapper` is creating a singleton of SchemaRegistryClient, when executing all tests SchemaRegistryClient is initialized only once by the first executing test and the SchemaRegistryClient will be using whatever was the credential setting for the first test and same is applied for all other tests, so executing test with and without auth is not feasible unless I makes changes to `RegistryClientWrapper` to reinitialize SchemaRegistryClient if there are any updates to url or auth. What do you think?

File: README.md (hunk 12 above)
Comment (gvishnutej) [reply README.md]: I would have to create a task and action to support  ```credentials {
  username = 'something'
  password = 'something else'
}``` correct?

Comment (ImFlog) [reply README.md]: You should be able to use extensions.
See https://docs.gradle.org/current/userguide/custom_plugins.html#sec:getting_input_from_the_build or https://docs.gradle.org/current/userguide/custom_plugins.html#sec:getting_input_from_the_build

File: src/test/kotlin/com/github/imflog/schema/registry/compatibility/CompatibilityTaskTest.kt (hunk 13 above)
Comment (ImFlog) [reply src/test/kotlin/com/github/imflog/schema/registry/compatibility/CompatibilityTaskTest.kt]: You are right about the fact that the registry client is a singleton but in the test context, we use this in the test itself :
```
val result: BuildResult? = GradleRunner.create()
                .withGradleVersion("4.9")
                .withProjectDir(folderRule.root)
                .withArguments(TEST_SCHEMAS_TASK)
                .withPluginClasspath()
                .withDebug(true)
                .build()
```
So the registry client is built for each test I think.

File: README.md (hunk 12 above)
Comment (gvishnutej) [reply README.md]: added support for 
```
credentials {
  username = 'something'
  password = 'something else'
}
```

File: src/test/kotlin/com/github/imflog/schema/registry/compatibility/CompatibilityTaskTest.kt (hunk 13 above)
Comment (gvishnutej) [reply src/test/kotlin/com/github/imflog/schema/registry/compatibility/CompatibilityTaskTest.kt]: I ran all test with debug, the `registryClient` variable in RegistryClientWrapper class is `null` for the first executing test. I made a change to re-initialize registryClient if the input config have changed

**Comment by ImFlog** [comment]
Sorry for the late answer. I will take a look at this during the week. 
I've spotted some small issues and if you don't mind I would like to commit on your branch. Could you open the maintainer edit Plz ? There is a button on the github PR which allows that.

**Comment by gvishnutej** [comment]
I have already selected the maintainer edit option, please feel free commit changes. Thanks

**Comment by ImFlog** [comment]
I successfully did all the change I wanted. Unfortunately, I can't find the doc about the way to launch a schema-registry with basic auth.
@gvishnutej, could you try and see if everything works as expected on a real use case ?
If it does, I'll merge this before doing some cleanup on the whole code base. Then I'll release a new version.

Thank you

**Comment by gvishnutej** [comment]
@ImFlog sure, I will validate and let you know. Thanks

**Comment by ImFlog** [comment]
@gvishnutej I am merging this as I want to work a bit on code cleanup of this repository.
If you find any bug, please tell me and I'll fix it :)
Thanks for your help on this !

**Pull request merged**